fh-android-sdk/src/main/java/org/json/fh/CDL.java

```diff
@@ -36,63 +36,79 @@ of this software and associated documentation files (the "Software"), to deal
  * delimited text is a very popular format for data interchange. It is
  * understood by most database, spreadsheet, and organizer programs.
  * <p>
- * Each row of text represents a row in a table or a data record. Each row ends with a NEWLINE character. Each row contains one or more values. Values are separated by commas. A value can contain any
+ * Each row of text represents a row in a table or a data record. Each row ends with a NEWLINE character.
+ * Each row contains one or more values. Values are separated by commas. A value can contain any
  * character except for comma, unless is is wrapped in single quotes or double quotes.
  * <p>
  * The first row usually contains the names of the columns.
  * <p>
- * A comma delimited list can be converted into a JSONArray of JSONObjects. The names for the elements in the JSONObjects can be taken from the names in the first row.
- * 
+ * A comma delimited list can be converted into a JSONArray of JSONObjects.
+ * The names for the elements in the JSONObjects can be taken from the names in the first row.
+ *
  * @author JSON.org
  * @version 2
  */
 public class CDL {
 
     /**
-     * Get the next value. The value can be wrapped in quotes. The value can
-     * be empty.
-     * 
+     * Gets the next value.
+     * The value can be wrapped in quotes. The value can be empty.
+     *
      * @param x A JSONTokener of the source text.
      * @return The value string, or null if empty.
-     * @throws JSONException this will be thrown if there is an error parsing the JSON if the quoted string is badly formed.
+     * @throws JSONException if there is an error parsing the JSON if the quoted string is badly formed.
      */
     private static String getValue(JSONTokener x) throws JSONException {
         char c;
+        char q;
+        StringBuffer sb;
         do {
             c = x.next();
-        } while (c <= ' ' && c != 0);
+        } while (c == ' ' || c == '\t');
         switch (c) {
-        case 0:
-            return null;
-        case '"':
-        case '\'':
-            return x.nextString(c);
-        case ',':
-            x.back();
-            return "";
-        default:
-            x.back();
-            return x.nextTo(',');
+            case 0:
+                return null;
+            case '"':
+            case '\'':
+                q = c;
+                sb = new StringBuffer();
+                while (true) {
+                    c = x.next();
+                    if (c == q) {
+                        break;
+                    }
+                    if (c == 0 || c == '\n' || c == '\r') {
+                        throw x.syntaxError("Missing close quote '" + q + "'.");
+                    }
+                    sb.append(c);
+                }
+                return sb.toString();
+            case ',':
+                x.back();
+                return "";
+            default:
+                x.back();
+                return x.nextTo(',');
```

Comment:
I would've made them so, but I didn't understand at first glance what they were supposed to be. I'll give it another sweep to see what I can do.